File: dotnet_rag/Services/DbService.cs
using System.Text.Json;
using Microsoft.SemanticKernel.Embeddings;
using Npgsql;
using Pgvector;

namespace PmRagChatbot.Services;

public class DbService
{
    private readonly NpgsqlDataSource _dataSource;
    private readonly ITextEmbeddingGenerationService _embeddingService;
    private readonly string _dataPath;

    public DbService(NpgsqlDataSource dataSource, ITextEmbeddingGenerationService embeddingService)
    {
        _dataSource = dataSource;
        _embeddingService = embeddingService;
        _dataPath = Path.Combine(AppContext.BaseDirectory, "data", "project_management_chunks.json");
    }

    public async Task SetupTableAsync()
    {
        await using var conn = await _dataSource.OpenConnectionAsync();

        await using var cmd = conn.CreateCommand();
        cmd.CommandText = "CREATE EXTENSION IF NOT EXISTS vector";
        await cmd.ExecuteNonQueryAsync();

        cmd.CommandText = """
                          CREATE TABLE IF NOT EXISTS pm_chunks (
                              id TEXT PRIMARY KEY,
                              chunk_text TEXT,
                              category TEXT,
                              embedding vector(1536)
                          )
                          """;
        await cmd.ExecuteNonQueryAsync();
    }

    public async Task SeedAsync()
    {
        await using var conn = await _dataSource.OpenConnectionAsync();

        await using var countCmd = conn.CreateCommand();
        countCmd.CommandText = "SELECT COUNT(*) FROM pm_chunks";
        var count = (long)(await countCmd.ExecuteScalarAsync() ?? 0);

        if (count > 0)
        {
            Console.WriteLine($"Already seeded ({count} records). Skipping.");
            return;
        }

        var json = await File.ReadAllTextAsync(_dataPath);
        var chunks = JsonSerializer.Deserialize<List<ChunkItem>>(json)!;

        Console.WriteLine($"Seeding {chunks.Count} chunks...");

        foreach (var chunk in chunks)
        {
            var chunkText = $"""
                             SORU: {chunk.Question}

                             CEVAP: {chunk.Answer}

                             BAĞLAM: {chunk.Context}
                             """;

            // ITextEmbeddingGenerationService — provider bağımsız
            var embeddingResult = await _embeddingService.GenerateEmbeddingAsync(chunkText);
            var vector = new Vector(embeddingResult.ToArray());

            await using var insertCmd = conn.CreateCommand();
            insertCmd.CommandText = """
                                    INSERT INTO pm_chunks (id, chunk_text, category, embedding)
                                    VALUES (@id, @chunk_text, @category, @embedding)
                                    ON CONFLICT (id) DO NOTHING
                                    """;
            insertCmd.Parameters.AddWithValue("id", chunk.Id);
            insertCmd.Parameters.AddWithValue("chunk_text", chunkText);
            insertCmd.Parameters.AddWithValue("category", chunk.Metadata.Category);
            insertCmd.Parameters.AddWithValue("embedding", vector);
            await insertCmd.ExecuteNonQueryAsync();
        }

        Console.WriteLine("Seeding complete.");
    }

    public async Task<float[]> GetEmbeddingAsync(string text)
    {
        var result = await _embeddingService.GenerateEmbeddingAsync(text);
        return result.ToArray();
    }

    public async Task<List<SearchResult>> SearchByEmbeddingAsync(float[] embedding, int topK = 3)
    {
        var vector = new Vector(embedding);
        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = conn.CreateCommand();

        cmd.CommandText = """
                          SELECT chunk_text, category,
                                 1 - (embedding <=> @embedding) AS similarity
                          FROM pm_chunks
                          ORDER BY embedding <=> @embedding
                          LIMIT @topK
                          """;
        cmd.Parameters.AddWithValue("embedding", vector);
        cmd.Parameters.AddWithValue("topK", topK);

        var results = new List<SearchResult>();
        await using var reader = await cmd.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            results.Add(new SearchResult(
                reader.GetString(0),
                reader.GetString(1),
                Math.Round(reader.GetDouble(2), 4)
            ));
        }

        return results;
    }
}

public record SearchResult(string ChunkText, string Category, double Similarity);

public record ChunkItem(
    string Id,
    string Question,
    string Answer,
    string Context,
    ChunkMetadata Metadata
);

public record ChunkMetadata(string Category);

File: python_rag/services/db_service.py
import json
from pathlib import Path
from sqlalchemy import create_engine, text
from openai import AzureOpenAI


class DbService:
    def __init__(self, connection_string: str, client: AzureOpenAI, embedding_deployment: str):
        self.engine = create_engine(connection_string)
        self.client = client
        self.embedding_deployment = embedding_deployment

    def setup_table(self):
        with self.engine.connect() as conn:
            conn.execute(text("CREATE EXTENSION IF NOT EXISTS vector"))
            conn.execute(text("""
                CREATE TABLE IF NOT EXISTS pm_chunks (
                    id TEXT PRIMARY KEY,
                    chunk_text TEXT,
                    category TEXT,
                    embedding vector(1536)
                )
            """))
            conn.commit()

    def get_embedding(self, text_input: str) -> list[float]:
        response = self.client.embeddings.create(
            input=text_input,
            model=self.embedding_deployment
        )
        return response.data[0].embedding

    def seed(self):
        data_path = Path(__file__).parent.parent / "data" / "project_management_chunks.json"
        chunks = json.loads(data_path.read_text(encoding="utf-8"))

        with self.engine.connect() as conn:
            count = conn.execute(text("SELECT COUNT(*) FROM pm_chunks")).scalar()
            if count > 0:
                print(f"Already seeded ({count} records). Skipping.")
                return

        print(f"Seeding {len(chunks)} chunks...")
        for chunk in chunks:
            chunk_text = f"""SORU: {chunk['question']}

CEVAP: {chunk['answer']}

BAĞLAM: {chunk['context']}"""

            embedding = self.get_embedding(chunk_text)

            with self.engine.connect() as conn:
                conn.execute(text("""
                    INSERT INTO pm_chunks (id, chunk_text, category, embedding)
                    VALUES (:id, :chunk_text, :category, CAST(:embedding AS vector))
                    ON CONFLICT (id) DO NOTHING
                """), {
                    "id": chunk["id"],
                    "chunk_text": chunk_text,
                    "category": chunk["metadata"]["category"],
                    "embedding": str(embedding)
                })
                conn.commit()

        print("Seeding complete.")

    def search_by_embedding(self, query_embedding: list[float], top_k: int = 3) -> list[dict]:
        embedding_str = str(query_embedding)
        with self.engine.connect() as conn:
            rows = conn.execute(text("""
                SELECT chunk_text, category,
                    1 - (embedding <=> CAST(:query_embedding AS vector)) AS similarity
                FROM pm_chunks
                ORDER BY embedding <=> CAST(:query_embedding AS vector)
                LIMIT :top_k
            """), {
                "query_embedding": embedding_str,
                "top_k": top_k
            }).fetchall()

        return [
            {
                "chunk_text": row.chunk_text,
                "category": row.category,
                "similarity": round(float(row.similarity), 4)
            }
            for row in rows
        ]